### DISCUSSION TITLE
Custom motion

### DISCUSSION THREAD

**Discussion by max397574** [folke/flash.nvim#24] [Show and tell]
This is something quite similar to leap-spooky.nvim (which I learned after I initially had the idea and started implementing it)
this allows you to use an operator (like `y`), then starts flash and allows you to go to any location where you can then enter another textobject or motion to be used with the initial operator and then it jumps back to the initial location

so you could e.g. do this `yr<keys with flash to jump>iw` to yank a word from any location and instantly jump back
as said this uses native operators and stuff so you can use all the textobject or motions available
```lua
vim.keymap.set("o", "r", function()
    local operator = vim.v.operator
    local register = vim--[[  ]].v.register
    vim.api.nvim_feedkeys(vim.api.nvim_replace_termcodes('<esc>', true, true, true), "o", true)
    vim.schedule(function()
        require("flash").jump({
            action = function(match, state)
                local op_func = vim.go.operatorfunc
                local saved_view = vim.fn.winsaveview()
                vim.api.nvim_set_current_win(match.win)
                vim.api.nvim_win_set_cursor(match.win, match.pos)
                _G.flash_op = function()
                    local start = vim.api.nvim_buf_get_mark(0, "[")
                    local finish = vim.api.nvim_buf_get_mark(0, "]")
                    vim.api.nvim_cmd({ cmd = "normal", bang = true, args = { "v" } }, {})
                    vim.api.nvim_win_set_cursor(0, { start[1], start[2] })
                    vim.cmd("normal! o")
                    vim.api.nvim_win_set_cursor(0, { finish[1], finish[2] })
                    vim.go.operatorfunc = op_func
                    vim.api.nvim_input('"' .. register .. operator)

                    vim.schedule(function()
                        vim.api.nvim_set_current_win(state.win)
                        vim.fn.winrestview(saved_view)
                    end)

                    _G.flash_op = nil
                end
                vim.go.operatorfunc = "v:lua.flash_op"
                vim.api.nvim_feedkeys("g@", "n", false)
            end,
        })
    end)
end)
```

**Comment by folke** [comment]
Nice!

**Comment by IndianBoy42** [comment]
Have you tried this with custom operators? I think you might have to restore `vim.go.operatorfunc` before the nvim_input to ensure it works (since you override operatorfunc to flash_op)

Also if I'm not wrong you can use `r` as the lhs. Atleast in base vim `r` has no meaning in operator-pending mode and it fits the mnemonic of 'remote'

**Reply by max397574** [reply]
makes sense
I choose `<c-w>` because it makes sense for personal reasons xD
but r sounds nice 👍 

updated

**Comment by folke** [comment]
Added to the examples!

**Comment by keevan** [comment]
@max397574 this is very magical, and is the first time I heard of this sort of movement. leap-spooky is such a great name too!

I came to point out that I tried the keymap described in the opening post, and had issues with `vim.keycode` being undefined. So I did some reading of the friendly manual and used an alternative which seems to work, which I hope helps anyone else who stumbles here.

### The small adjustment
`vim.keycode('<esc>')` -> `vim.api.nvim_replace_termcodes('<esc>', true, true, true)`

I also found it unclear initially, given the example presented `y<c-w><keys with flash to jump>iw` since it mentions `<c-w>`, and after more reading the `<c-w>` was your personal binding, although the `r` was used in the example code instead. Could we update the example too?

Thanks again for sharing the keymap!

**Reply by max397574** [reply]
vim.keycode is just a new thing that was added on nightly
I'll adjust to the old function which does the same

**Comment by folke** [comment]
I just added this as a new plugin.

Use it like this:
```lua
vim.keymap.set("o", "r", function()
  require("flash").remote()
end)
```

Full code here https://github.com/folke/flash.nvim/blob/main/lua/flash/plugins/remote.lua

Thanks again! Really cool!

It's really useful in combination with flash treesitter etc. Love it!

**Reply by folke** [reply]
Will add it to the docs later

**Reply by IndianBoy42** [reply]
Correct me if I'm wrong but the current implementation assumes the operator works in visual mode which may not be true for custom operators right?

**Reply by dpetka2001** [reply]
When I try to use it as plugin from flash with `yr` no labels are being shown (after this has been added as a plugin to flash by @folke), but if I set the code in the first post as a keymap directly to `keymaps.lua` (in LazyVim) everything works correctly. Is there some difference between the 2? Am I not calling it the correct way as a flash plugin?

**Reply by folke** [reply]
Did you update flash/LazyVim?

It works for me and I also didn't change the logic.
- `yr` starts jump mode
- enter some characters and pick a label
- your cursor will now be at the label position waiting for an motion
- you could do `iw` or start treesitter mode with `S` for example.

Can you clarify what isn't working specifically?

**Reply by dpetka2001** [reply]
When I press `yr` jump mode remote does not start. If i start typing no labels are shown. Everything is interpreted as normal vim commands. So if i type `l` (lowercase L) it just moves the cursor from its current position to the right. If I set the code in the first post as a keymap in `keymaps.lua` everything works as expected like you also described. If i press `y` (for yank operator) and then type `:lua require("flash").remote()` it also works as expected. I can see the jump mode taking effect.

**Reply by folke** [reply]
I didn't make a new LazyVim release yet and you're probably following stable LazyVim releases.
Just pushed a new release, so should work in a couple of minutes.

**Reply by folke** [reply]
New LazyVim release is available!

**Reply by dpetka2001** [reply]
Oh I'm so sorry for bothering you. I thought it was already in the LazyVim. Yes with the new release everything works correctly. So sorry again.

**Reply by folke** [reply]
My bad, forgot to create a new release :)

**Reply by dpetka2001** [reply]
I also read about it in the Lazyvim site docs and thought it was already integrated. Didn't check the releases to see if it was already merged or not.

**Reply by folke** [reply]
If you'd rather follow dev releases, check https://www.lazyvim.org/configuration/lazy.nvim

**Reply by dpetka2001** [reply]
Yes, I'm aware of that. Thank you very much again for your help.

**Comment by folke** [comment]
I just updated the remote plugin to better deal with the `c` operator.

When doing `cr` you will now jump back when leaving insert mode in the remote location.

That's actually really useful to quickly make a remote edit and get back to where you came from.

**Comment by al-ce** [comment]
This is very cool, thanks @max397574 and @folke

I'm throwing this together before heading out for the day, but I was curious about imitating leap-spooky's 'paste on remote yank' feature (for entire lines or standard objects). First, I made a custom text object for the current line from the info in this [stack exchange](https://vi.stackexchange.com/questions/6101/is-there-a-text-object-for-current-line) post. Then I added this to `flash.plugins.remote.op`, just stealing from the recent commit for the `c` operator 10750139d3d4f2fb6c7bb8cc33aef988a7b26b7c

```lua
  if M.operator == "c" then
    vim.api.nvim_create_autocmd("InsertLeave", {
      once = true,
      callback = M.restore,
    })
  elseif M.operator == "y" then -- need to check for some opt to paste after yank
    vim.api.nvim_create_autocmd("TextYankPost", {
      once = true,
      callback = function()
        vim.schedule(function()
          M.restore()
          -- but should also distinguish pasting on a new line and at cursor
          vim.api.nvim_input("o<esc>p")
        end)
      end,
    })
  else
    vim.schedule(M.restore)
  end
```

This is the result (pardon the highlighting issues w/ the params)

https://github.com/folke/flash.nvim/assets/23170004/517ec52f-4144-4fff-8e7b-60c9464aa0d4

I want to set some option to distinguish between immediately pasting post yank and not, and also whether to paste on a new line or at the cursor (since I only really want to go to a new line if I'm yanking an entire line). Just messing around a bit and thought I'd share, I know this is overkill for saving one to two keystrokes.

**Reply by folke** [reply]
Nice! But yes, I much more prefer to use a paste afterwards myself :)

But doing the restore on TextYankPost might be a good idea. I've noticed some weirdness regarding yanking sometimes. Still haven't figured out what exactly is causing it and when.

**Reply by IndianBoy42** [reply]
Maybe we can get an `on_restore` callback to let the user create things like this?

**Comment by DeepReef11** [comment]
Is remote paste available?
